src/utils/changes.ts

```diff
@@ -175,3 +181,152 @@ export function prependChangeset(
+  title?: string;
+  description?: string;
+  state?: string;
+  prs: string[];
+}
+
+// This is set to 8 since GitHub and GitLab prefer that over the default 7 to
+// avoid collisions.
+const SHORT_SHA_LENGTH = 8;
+function formatCommit(commit: Commit): string {
+  let text = `- ${commit.title}`;
+  if (!commit.hasPRinTitle) {
+    const link = commit.pr
+      ? `#${commit.pr}`
+      : commit.hash.slice(0, SHORT_SHA_LENGTH);
+    text = `${text} (${link})`;
+  }
+  return text;
+}
+
+export async function generateChangesetFromGit(
+  git: SimpleGit,
+  rev: string
+): Promise<string> {
+  const { repo, owner } = await getGlobalGithubConfig();
+  const gitCommits = await getChangesSince(git, rev);
+  const commitsQuery = gitCommits
+    .map(({ hash }) => `C${hash}: object(oid: "${hash}") {...PRFragment}`)
+    .join('\n');
+
+  const github = getGithubClient();
+  const { repository: githubCommits } = await github.graphql(`{
+    repository(name: "${repo}", owner: "${owner}") {
+      ${commitsQuery}
+    }
+  }
+
+  fragment PRFragment on Commit {
+    associatedPullRequests(first: 1) {
+      nodes {
+        number
+        milestone {
+          number
+        }
+      }
+    }
+  }`);
+
+  const milestones: {
+    [number: string]: Milestone;
+  } = {};
+  const commits: {
+    [hash: string]: Commit;
+  } = {};
+  const leftovers: Commit[] = [];
+  const missing: Commit[] = [];
+  for (const gitCommit of gitCommits) {
+    const hash = gitCommit.hash;
+
+    const commit = (commits[hash] = {
+      hash: hash,
+      title: gitCommit.message,
+      pr: gitCommit.pr,
+      hasPRinTitle: Boolean(gitCommit.pr),
+      milestone: null,
+    });
+    const commitInfo = githubCommits[`C${hash}`];
+    if (commitInfo) {
+      const pr = commitInfo.associatedPullRequests.nodes[0];
+      if (pr) {
+        commit.pr = pr.number;
+        commit.milestone = pr.milestone?.number;
+      }
+    } else {
+      missing.push(commit);
+    }
+
+    const milestoneNum = commit.milestone;
+    if (milestoneNum != null) {
+      const milestone = milestones[milestoneNum] || { prs: [] };
+      // We _know_ the PR exists as milestones are attached to PRs
+      milestone.prs.push(commit.pr as string);
+      milestones[milestoneNum] = milestone;
+    } else {
+      leftovers.push(commit);
+    }
+  }
+
+  if (missing.length > 0) {
+    logger.warn(
+      `The following commits were not found on GitHub:`,
+      missing.map(commit => `${commit.hash.slice(0, 8)} ${commit.title}`)
+    );
+  }
+
+  const milestoneQuery = Object.keys(milestones)
+    .map(
+      number =>
+        `M${number}: milestone(number: ${number}) { ...MilestoneFragment }`
+    )
+    .join('\n');
+
+  const { repository: milestonesInfo } = await github.graphql(`{
+    repository(name: "${repo}", owner: "${owner}") {
+      ${milestoneQuery}
+    }
+  }
+
+    fragment MilestoneFragment on Milestone {
+      title
+      description
+      state
+    }
+  `);
```

Comment:
There are some GraphQL queries. What do you think about encapsulating them in methods and calling these methods, instead of doing them in here? I suggest placing them in `src/utils/githubApi`.

Reply:
These are very specific queries so I wouldn't put them into the very generic `src/utils/githubApi` but I thought about moving them into separate functions. Just held that off until I got some feedback from you folks to support that idea. Will do 😄